Documentation/design-docs/tailcalls-with-helpers.md

```diff
@@ -0,0 +1,439 @@
+The thunk will now extract parameters for D from the thread local storage and call D.
+```
+* Return address of A
+* Callee saved registers and locals of A
+* Stack arguments of B
+* Return address of B
+* Callee saved registers and locals of CallTarget thunk for D
+```
+### In D
+We are in the last function of the chain, so after it does its work, it returns to its CallTarget thunk. 
+```
+* Return address of A
+* Callee saved registers and locals of A
+* Stack arguments of B
+* Return address of B
+* Callee saved registers and locals of CallTarget thunk for D
+* Stack arguments of D
+* Return address of D
+* Callee saved registers and locals of D
+```
+### Returned to CallTarget thunk for D with ChainCall flag in the TailCallHelperStackEntry **clear**
+The thunk checks the ChainCall flag and since it is clear, it recognizes we are now returning from the call chain and so it returns the result of the D. 
+```
+* Return address of A
+* Callee saved registers and locals of A
+* Stack arguments of B
+* Return address of B
+* Callee saved registers and locals of CallTarget thunk for D
+```
+### Returned to A
+We are back in A and we have the return value of the call chain.
+```
+* Return address of A
+* Callee saved registers and locals of A
+```
+
+## Example of thunks generated for a simple generic method
+
+```C#
+struct Point
+{
+    public int x;
+    public int y;
+    public int z;
+}
+
+class Foo
+{
+    public Point Test<T>(int x, T t) where T : class
+    {
+        Console.WriteLine("T: {0}", typeof(t));
+        return new Point(x, x, x);
+    }
+}
+```
+
+For tail calling the Test function, the IL helpers could look e.g. as follows:
+
+```IL
+.method public static void StoreArgumentsTest(native int target, object thisObj, native int genericContext, int x, object t) cil managed
+{
+    .maxstack 4
+
+    call native int FetchArgumentBuffer()
+
+    dup
+    ldftn Point CallTargetTest()
+    stind.i
+    sizeof native int
+    add
+
+    dup
+    ldarg.1 // "thisObj"
+    stobj object
+    sizeof object
+    add
+
+    dup
+    ldarg.2 // "genericContext"
+    stind.i
+    sizeof native int
+    add
+
+    dup
+    ldarg.3 // "x"
+    stind.i4
+    sizeof native int
+    add
+
+    dup
+    ldarg.4 // "t"
+    stobj object
+    sizeof object
+    add
+
+    ldarg.0 // "target"
+    stind.i
+
+    ret
+}
+
+```
+
+```IL
+.method public static Point CallTargetTest() cil managed
+{
+    .maxstack 4
+    .locals init (
+       [0] native int buffer,
+       [1] valuetype TailCallHelperStackEntry entry,
+       [2] Point result
+    )
+
+    // Initialize the TailCallHelperStackEntry
+    // chainCall = false
+    // sp = current sp
+    // next = t_helperStackTOS
+    ldloca.s 1
+    ldc.i4.0
+    stfld bool TailCallHelperStackEntry::chainCall
+    ldloca.s 1
+    call native int GetCurrentSp()
+    stfld native int TailCallHelperStackEntry::sp
+    ldloca.s 1
+    ldfld native int t_helperStackTOS
```

Comment:
@jkotas You mean the t_helperStackTOS? Yes, I agree. I have put it this way into the example thinking it would be more obvious how it works. I'll change the example to match the reality.